File: ios/InAppPurchaseSample/InAppPurchaseSample/ViewController.m
//
//  ViewController.m
//  InAppPurchaseSample
//
//  Created by fujita taisuke on 2012/10/26.
//  Copyright (c) 2012年 Revolution 9. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
}

- (IBAction)respondToBuyBottonClick:(id)sender
{
    if (![SKPaymentQueue canMakePayments]) {
        UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"エラー"
                                                        message:@"アプリ内課金の使用が制限されています。"
                                                       delegate:nil
                                              cancelButtonTitle:nil
                                              otherButtonTitles:@"OK", nil];
        [alert show];
        return;
    }
    NSSet *set = [NSSet setWithObjects:@"com.9revolution9.adsremover", nil];
    SKProductsRequest *productsRequest = [[SKProductsRequest alloc] initWithProductIdentifiers:set];
    productsRequest.delegate = self;
    [productsRequest start];
}

- (IBAction)respondToRestoreBottonClick:(id)sender
{
    [[SKPaymentQueue defaultQueue] restoreCompletedTransactions];
}

- (void)alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)buttonIndex {
    // 途中で止まった処理を再開する Consumable アイテムにも有効
    [[SKPaymentQueue defaultQueue] restoreCompletedTransactions];
}

- (void)productsRequest:(SKProductsRequest *)request didReceiveResponse:(SKProductsResponse *)response {
    // 無効なアイテムがないかチェック
    if ([response.invalidProductIdentifiers count] > 0) {
        UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"エラー"
                                                        message:@"アイテムIDが不正です。"
                                                       delegate:nil
                                              cancelButtonTitle:@"OK"
                                              otherButtonTitles:nil, nil];
        [alert show];
        return;
    }
    // 購入処理開始
    [[SKPaymentQueue defaultQueue] addTransactionObserver:self];
    for (SKProduct *product in response.products) {
        SKPayment *payment = [SKPayment paymentWithProduct:product];
        [[SKPaymentQueue defaultQueue] addPayment:payment];
    }
}

- (void)applicationDidEnterBackground
{
    // リストア処理開始
    [[SKPaymentQueue defaultQueue] removeTransactionObserver:self];
}

- (void)applicationWillEnterForeground
{
    [[SKPaymentQueue defaultQueue] addTransactionObserver:self];
}

- (void)paymentQueue:(SKPaymentQueue *)queue updatedTransactions:(NSArray *)transactions {
    for (SKPaymentTransaction *transaction in transactions) {
        if (transaction.transactionState == SKPaymentTransactionStatePurchasing) {
             // 購入処理中
        } else if (transaction.transactionState == SKPaymentTransactionStatePurchased) {
            [queue finishTransaction:transaction];
        } else if (transaction.transactionState == SKPaymentTransactionStateFailed) {
            [queue finishTransaction:transaction];
            if (transaction.error.code == SKErrorUnknown) {
                UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"処理が途中中断されました"
                                                                message:[transaction.error localizedDescription]
                                                               delegate:self
                                                      cancelButtonTitle:nil
                                                      otherButtonTitles:@"OK", nil];
                [alert show];
            }
        } else {
            // 復帰処理完了
            [queue finishTransaction:transaction];
        }
    }
}

- (void)paymentQueue:(SKPaymentQueue *)queue removedTransactions:(NSArray *)transactions
{
    for (SKPaymentTransaction *transaction in transactions) {
        if (transaction.transactionState == SKPaymentTransactionStatePurchased) {
            self.label.text = @"購入しました。";
        }
    }
    [[SKPaymentQueue defaultQueue] removeTransactionObserver:self];
}

- (void)paymentQueue:(SKPaymentQueue *)queue restoreCompletedTransactionsFailedWithError:(NSError *)error
{
    if ([queue.transactions count] == 0) {
        [[SKPaymentQueue defaultQueue] removeTransactionObserver:self];
    }
}

- (void)paymentQueueRestoreCompletedTransactionsFinished:(SKPaymentQueue *)queue
{
    for (SKPaymentTransaction *transaction in queue.transactions) {
        if (transaction.transactionState == SKPaymentTransactionStatePurchased) {
            self.label.text = @"リストアしました。";
        }
    }
    if ([queue.transactions count] == 0) {
        [[SKPaymentQueue defaultQueue] removeTransactionObserver:self];
    }
}

@end

File: ios_swift/InAppPurchase/InAppPurchase/ViewController.swift
//
//  ViewController.swift
//  InAppPurchase
//
//  Created by taisuke fujita on 2014/10/11.
//  Copyright (c) 2014年 taisuke fujita. All rights reserved.
//

import UIKit
import StoreKit

class ViewController: UIViewController, SKProductsRequestDelegate, SKPaymentTransactionObserver {
    
    @IBOutlet weak var label: UILabel!
    
    func productsRequest(request: SKProductsRequest!,
        didReceiveResponse response: SKProductsResponse!) {
            if response.invalidProductIdentifiers.count > 0 {
                let alertController = UIAlertController(title: "課金アイテムチェック",
                    message: "アイテムIDが不正です。",
                    preferredStyle: .Alert)
                let action = UIAlertAction(title: "閉じる", style: .Default, handler: nil)
                alertController.addAction(action)
                self.presentViewController(alertController, animated: true, completion: nil)
                return
            }
            SKPaymentQueue.defaultQueue().addTransactionObserver(self)
            for product in response.products {
                let payment = SKPayment(product: product as! SKProduct)
                SKPaymentQueue.defaultQueue().addPayment(payment)
            }
    }
    
    func paymentQueue(queue: SKPaymentQueue!, updatedTransactions transactions: [AnyObject]!) {
        for transaction in transactions as! [SKPaymentTransaction] {
            switch transaction.transactionState {
            case .Purchasing:
                println("処理中...")
            case .Purchased:
                let url = NSBundle.mainBundle().appStoreReceiptURL
                let data = NSData(contentsOfURL: url!)!
                let base64 = data.base64EncodedStringWithOptions(NSDataBase64EncodingOptions(rawValue:0))
                println(base64)
                queue.finishTransaction(transaction)
            case .Restored:
                queue.finishTransaction(transaction)
            case .Failed:
                println(transaction.error.localizedDescription)
                queue.finishTransaction(transaction)
                if transaction.error.code == SKErrorUnknown {
                    // リストアする
                    SKPaymentQueue.defaultQueue().restoreCompletedTransactions()
                }
                
            default:
                println()
            }
        }
    }
    
    func paymentQueue(queue: SKPaymentQueue!, removedTransactions transactions: [AnyObject]!) {
        for transaction in transactions as! [SKPaymentTransaction] {
            if transaction.transactionState == .Purchased {
                self.label.text = "\(transaction.payment.productIdentifier)"
            }
        }
        SKPaymentQueue.defaultQueue().removeTransactionObserver(self)
    }
    
    func paymentQueue(queue: SKPaymentQueue!, restoreCompletedTransactionsFailedWithError error: NSError!) {
        if queue.transactions.count == 0 {
            SKPaymentQueue.defaultQueue().removeTransactionObserver(self)
        }
    }
    
    func paymentQueueRestoreCompletedTransactionsFinished(queue: SKPaymentQueue!) {
        for transaction in queue.transactions as! [SKPaymentTransaction] {
            if transaction.transactionState == .Purchased {
                self.label.text = "\(transaction.payment.productIdentifier)"
            }
        }
        if queue.transactions.count == 0 {
            SKPaymentQueue.defaultQueue().removeTransactionObserver(self)
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        if !SKPaymentQueue.canMakePayments() {
            // エラーメッセージを表示する
            let alertController = UIAlertController(title: "課金設定チェック",
                message: "アプリ内課金の使用が制限されています。",
                preferredStyle: .Alert)
            
            let action = UIAlertAction(title: "閉じる", style: .Default, handler: nil)
            alertController.addAction(action)
            self.presentViewController(alertController, animated: true, completion: nil)
            return
        }
        let center = NSNotificationCenter.defaultCenter()
        center.addObserver(self, selector: "applicationDidEnterBackground",
            name: UIApplicationDidEnterBackgroundNotification, object: nil)
        center.addObserver(self, selector: "applicationWillEnterForeground",
            name: UIApplicationWillEnterForegroundNotification, object: nil)
    }
    
    func applicationDidEnterBackground() {
        SKPaymentQueue.defaultQueue().removeTransactionObserver(self)
    }

    func applicationWillEnterForeground() {
        SKPaymentQueue.defaultQueue().addTransactionObserver(self)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    @IBAction func respondToBuyButtonClick() {
        let set = NSSet(objects: "com.9revolution9.removeads")
        let productRequest = SKProductsRequest(productIdentifiers: set as Set<NSObject>)
        productRequest.delegate = self
        productRequest.start()
    }
    
    @IBAction func respondToRestoreButtonClick() {
        SKPaymentQueue.defaultQueue().restoreCompletedTransactions()
    }
}

